### ISSUE TITLE
Build fails unpredictably if _CoqProject file contains missing a file

### ISSUE THREAD

**Issue by anton-trunov** [coq/coq#9319]
#### Version

8.10+alpha(d58614c1adb94fbfbb75e385bb75f768022b9629)

#### Operating system

macOS 10.14.2 & Debian Stretch

#### Description of the problem

E.g. this

```shell
touch a.v
echo 'Require a.' > b.v
printf "a.v\nb.v\n\no_such_file.v" > _CoqProject
coq_makefile -f _CoqProject -o Makefile
make VERBOSE=1
```

produces the following output:

```shell
/Applications/Xcode.app/Contents/Developer/usr/bin/make --no-print-directory -f "Makefile" pre-all
if [ "8.10+alpha" != "8.10+alpha" ]; then\
	  echo "W: This Makefile was generated by Coq 8.10+alpha";\
	  echo "W: while the current Coq version is 8.10+alpha";\
	fi
/Applications/Xcode.app/Contents/Developer/usr/bin/make --no-print-directory -f "Makefile" real-all
make: *** [all] Error 2
```

So it does not provide any pointers as to where to look for the source of error.

Also (this happened on a real project), when there is a missing file (basically a typo in the file name) in `_CoqProject` file, then some of dependencies get resolved incorrectly and the build is failing at an unrelated moment (too early to quickly understand what has caused it).

**Comment by gares** [comment]
I can't reproduce it:
```shell
$ coq_makefile foobar.v -o xxx
$ make -f xxx
make[1]: *** No rule to make target 'foobar.v', needed by 'foobar.vo'.  Stop.
xxx:312: recipe for target 'all' failed
```

what is `/Applications/Xcode.app/Contents/Developer/usr/bin/make` ?

**Comment by anton-trunov** [comment]
I updated my Coq setup to the most recent commit and I get a bit different output now (please see the updated original post).

> what is /Applications/Xcode.app/Contents/Developer/usr/bin/make ?

```shell
$ /Applications/Xcode.app/Contents/Developer/usr/bin/make --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```

**Comment by anton-trunov** [comment]
Here is another way to reproduce the problem:

```shell
$ git clone https://github.com/imdea-software/fcsl-pcm
$ cd fcsl-pcm
$ echo "foobar.v" >> _CoqProject
$ make
coq_makefile -f _CoqProject -o CoqMakefile
/Applications/Xcode.app/Contents/Developer/usr/bin/make --no-print-directory -f CoqMakefile
COQC finmap/finmap.v
File "./finmap/finmap.v", line 21, characters 25-32:
Error: Unable to locate library ordtype with prefix fcsl.

make[2]: *** [finmap/finmap.vo] Error 1
make[1]: *** [all] Error 2
make: *** [invoke-coqmakefile] Error 2
```

`finmap/finmap.v` depends on `finmap/ordtype.v`, but because of that extra `foobar.v`file  the dependency gets lost and the compilation error is confusing.

**Comment by SkySkimmer** [comment]
Looks like an `-include` issue sort of like https://github.com/mit-plv/fiat/pull/21
We do `-include $(ALLDFILES)` which contains `$(VDFILE)`, and `$(VDFILE) : $(VFILES)` can't be built when a vfile is missing but since it's `-include` there is no error.

**Comment by SkySkimmer** [comment]
See #9324

**Comment by RalfJung** [comment]
@gares try `rm .coqdeps.d` before doing `make` again. The deps-file will not be rebuilt because of the missing file, but the build still continues -- without dependency information and hence in a random order.

**Comment by RalfJung** [comment]
A partial fix is to add `$(VDFILE).d` as an extra dependency of the `$(VOFILES)` rule.

**Comment by Ptival** [comment]
Oof, I just spent a lot of time trying to figure out why `coq_makefile` would suddenly not generate correct dependencies!

This is very unfortunate, a warning would be greatly appreciated!

**Comment by JasonGross** [comment]
> A partial fix is to add `$(VDFILE).d` as an extra dependency of the `$(VOFILES)` rule.

This is a very bad idea.  It means that *all* .vo files will be rebuilt any time you add any new .v file or change any import in any file.  Maybe an order-only dependency will work (I've forgotten exactly how those work)?

**Comment by JasonGross** [comment]
Yes, I think [order-only prerequisites](https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html) are what you want

**Comment by fakusb** [comment]
The original example does dail wirh a `good` message during `make`, but the problem still occurs in the sense that if a file is missing, coqdep does not give a warning or error, but just generates wrong order of dependencies. 

````
touch a.v
echo 'Require a.' > b.v
printf "b.v\na.v\nno_such_file.v\n" > _CoqProject
coq_makefile -f _CoqProject -o Makefile
make VERBOSE=1
````
(a.v and b.v switched compared to original example)
I think I will try to add that coqdep prints to STDERROR if a file is missing

**Comment by gares** [comment]
good idea

**Comment by fakusb** [comment]
For the record: If the makefile would just build VDFILE in the `all` target, there would be an nice error message.

So calling `make .Makefile.d` can already be used now to find the file in _CoqProject without matching .v file.
